File: tools/clicker.go
package main

import (
    "context"
    "encoding/json"
    "fmt"
    "io"
    "net"
    "net/http"
    "time"
    "github.com/chromedp/chromedp"
)

const (
    targetURL        = "chrome://transfer-frame/"
    browserDebugPort = 9222
    maxWaitTime      = 30 * time.Second
    retryInterval    = 500 * time.Millisecond
)

func main() {
    fmt.Println("[*] 嘗試連接 Chrome remote debugging...")

    // 1️⃣ 等待端口
    if err := waitPort("127.0.0.1", browserDebugPort, maxWaitTime); err != nil {
        fmt.Println(err)
        return
    }
    fmt.Printf("[✓] 端口 %d 已啟動\n", browserDebugPort)

    // 2️⃣ 取得 WebSocket URL
    wsURL, err := waitWebSocketURL(maxWaitTime)
    if err != nil {
        fmt.Println(err)
        return
    }
    fmt.Println("[✓] WebSocket URL:", wsURL)

    // 3️⃣ 建立 chromedp context
    allocCtx, allocCancel := chromedp.NewRemoteAllocator(context.Background(), wsURL)
    defer allocCancel()

    ctx, ctxCancel := chromedp.NewContext(allocCtx)
    defer ctxCancel()

    ctx, timeoutCancel := context.WithTimeout(ctx, 20*time.Second)
    defer timeoutCancel()

    // 4️⃣ 執行任務
    fmt.Println("[*] 開始導航到:", targetURL)
    tasks := chromedp.Tasks{
        chromedp.Navigate(targetURL),
        chromedp.WaitVisible(`#js-start_all_download`, chromedp.ByID),
        chromedp.Click(`#js-start_all_download`, chromedp.ByID),
    }

    fmt.Println("[*] 開始執行任務...")
    if err := chromedp.Run(ctx, tasks); err != nil {
        fmt.Printf("[x] 操作失敗: %v\n", err)
        return
    }

    fmt.Println("[✓] 按鈕已成功觸發！")
}

// waitPort 持續嘗試 TCP 連線，最多等待 timeout
func waitPort(host string, port int, timeout time.Duration) error {
    deadline := time.Now().Add(timeout)
    addr := fmt.Sprintf("%s:%d", host, port)
    attempt := 0

    for time.Now().Before(deadline) {
        attempt++
        conn, err := net.DialTimeout("tcp", addr, 1*time.Second)
        if err == nil {
            conn.Close()
            fmt.Printf("[✓] 第 %d 次嘗試: 連線成功\n", attempt)
            return nil
        }
        fmt.Printf("[*] 第 %d 次嘗試: 連線失敗 (%v)\n", attempt, err)
        time.Sleep(retryInterval)
    }
    return fmt.Errorf("[x] 等待端口 %s 超時", addr)
}

// waitWebSocketURL 持續嘗試取得 WebSocket URL，最多等待 timeout
func waitWebSocketURL(timeout time.Duration) (string, error) {
    deadline := time.Now().Add(timeout)
    url := fmt.Sprintf("http://127.0.0.1:%d/json/version", browserDebugPort)
    attempt := 0

    for time.Now().Before(deadline) {
        attempt++
        resp, err := http.Get(url)
        if err == nil {
            body, err := io.ReadAll(resp.Body)
            resp.Body.Close()
            if err == nil {
                var v struct {
                    WebSocketDebuggerUrl string `json:"webSocketDebuggerUrl"`
                }
                if err := json.Unmarshal(body, &v); err == nil && v.WebSocketDebuggerUrl != "" {
                    fmt.Printf("[✓] 第 %d 次嘗試: WebSocket URL 已取得\n", attempt)
                    return v.WebSocketDebuggerUrl, nil
                }
            }
        }
        fmt.Printf("[*] 第 %d 次嘗試: WebSocket 尚未就緒 (%v)\n", attempt, err)
        time.Sleep(retryInterval)
    }

    return "", fmt.Errorf("[x] 等待 WebSocket URL 超時")
}

File: tools/clicker.js
const puppeteer = require("puppeteer-core");
const http = require("http");
const net = require("net");

const browserDebugHost = "127.0.0.1";
const browserDebugPort = 9222;
const targetURL = "chrome://transfer-frame/";
const maxWaitTime = 30 * 1000;
const retryInterval = 500;

async function waitPort(host, port, timeout) {
  const start = Date.now();
  let attempt = 0;

  while (Date.now() - start < timeout) {
    attempt++;
    try {
      await new Promise((resolve, reject) => {
        const socket = net.createConnection({ host, port });
        socket.once("connect", () => {
          socket.end();
          resolve();
        });
        socket.once("error", reject);
      });
      console.log(`[✓] 第 ${attempt} 次嘗試: 連線成功`);
      return;
    } catch (err) {
      console.log(`[*] 第 ${attempt} 次嘗試: 連線失敗 (${err.message})`);
      await new Promise((r) => setTimeout(r, retryInterval));
    }
  }

  throw new Error(`[x] 等待端口 ${host}:${port} 超時`);
}

// 等待 WebSocket URL
async function waitWebSocketURL(timeout) {
  const start = Date.now();
  const url = `http://${browserDebugHost}:${browserDebugPort}/json/version`;
  let attempt = 0;

  while (Date.now() - start < timeout) {
    attempt++;
    try {
      const wsURL = await new Promise((resolve, reject) => {
        http
          .get(url, (res) => {
            let data = "";
            res.on("data", (chunk) => (data += chunk));
            res.on("end", () => {
              try {
                const parsed = JSON.parse(data);
                if (parsed.webSocketDebuggerUrl)
                  resolve(parsed.webSocketDebuggerUrl);
                else reject("webSocketDebuggerUrl not found");
              } catch (err) {
                reject(err);
              }
            });
          })
          .on("error", reject);
      });
      console.log(`[✓] 第 ${attempt} 次嘗試: WebSocket URL 已取得`);
      return wsURL;
    } catch (err) {
      console.log(`[*] 第 ${attempt} 次嘗試: WebSocket 尚未就緒 (${err})`);
      await new Promise((r) => setTimeout(r, retryInterval));
    }
  }

  throw new Error("[x] 等待 WebSocket URL 超時");
}

(async () => {
  console.log("[*] 嘗試連接 Chrome remote debugging...");

  let browser, page;

  try {
    await waitPort(browserDebugHost, browserDebugPort, maxWaitTime);
    console.log(`[✓] 端口 ${browserDebugPort} 已啟動`);

    const wsURL = await waitWebSocketURL(maxWaitTime);
    console.log("[✓] WebSocket URL:", wsURL);

    browser = await puppeteer.connect({
      browserWSEndpoint: wsURL,
      defaultViewport: null,
    });
    page = await browser.newPage();

    console.log("[*] 開始導航到:", targetURL);
    await page.goto(targetURL, { waitUntil: "load" });

    console.log("[*] 等待列表渲染...");
    try {
      await page.waitForSelector(
        "div.download-list#js-download-box > ul > li",
        {
          timeout: 10000,
        }
      );
      console.log("[*] 列表已渲染完成");
    } catch (err) {
      console.error("[x] 列表渲染異常:", err);
      console.log("[*] 列表未渲染，略過等待");
    }

    console.log("[*] 等待下載按鈕渲染...");
    try {
      await page.waitForSelector("#js-start_all_download", {
        visible: true,
        timeout: 10000,
      });
      console.log("[*] 下載按鈕已渲染完成");
    } catch (err) {
      console.error("[x] 下載按鈕渲染異常:", err);
      console.log("[*] 下載按鈕未渲染，略過等待");
    }

    await page.click("#js-start_all_download");
    console.log("[✓] 下載按鈕已觸發！");
  } catch (err) {
    console.error("[x] 操作失敗:", err);
  } finally {
    if (page) await page.close();
    if (browser) await browser.disconnect();
    process.exit(0);
  }
})();
